src/TableFunctions/TableFunctionFile.cpp

```diff
@@ -7,10 +7,92 @@
 #include <Storages/ColumnsDescription.h>
 #include <Storages/StorageFile.h>
 #include <TableFunctions/TableFunctionFactory.h>
+#include <Interpreters/evaluateConstantExpression.h>
+#include <Formats/FormatFactory.h>
+#include <Parsers/ASTIdentifier_fwd.h>
 
 namespace DB
 {
 
+namespace ErrorCodes
+{
+    extern const int UNKNOWN_IDENTIFIER;
+    extern const int LOGICAL_ERROR;
+    extern const int NUMBER_OF_ARGUMENTS_DOESNT_MATCH;
+    extern const int BAD_ARGUMENTS;
+}
+
+void TableFunctionFile::parseArguments(const ASTPtr & ast_function, ContextPtr context)
```

Comment:
There is a lot of unnecessary code duplication from ITableFunctionFileLike::parseArguments. I would prefer just to add two new virtual methods in ITableFunctionFileLike.
1 - parseFirstArgument(...). Default implementation - extracting String literal, TableFunctionFile will have it's own implementation that takes into account that argument could be file descriptor
2 - getFormatFromFirstArgument(...). Default implementation - getFormatFromFileName, TableFunctionFile will also try getFormatFromFileDescriptor in case of file descriptor.

Also we want to support file descriptors only for clickhouse-local, not for server. So, in parseFirstArgument in TableFunctionFile we should check that we are run as local tool (like it's checked [here](https://github.com/ClickHouse/ClickHouse/blob/ca7ba0a19878aefab1d7de649df5a6a519fc660e/src/Storages/StorageFile.cpp#L183).

Also current implementation doesn't take into account case with file descriptor in method getActualTableStructure when structure is 'auto'. Although it's not obvious how to implement it with file descriptor properly, so, by now we can just throw an exception that we don't support schema inference for table function file with file descriptors and then we can improve this place.

Reply:
ok, i try to implement it.